src/main/scala/scalismo/common/Scalar.scala

```diff
@@ -0,0 +1,283 @@
+    override def toFloat(a: UInt): Float = a.toFloat
+    override def toDouble(a: UInt): Double = a.toDouble
+
+    override def fromByte(n: Byte): UInt = UInt(n.toInt)
+    override def fromShort(n: Short): UInt = UInt(n.toInt)
+    override def fromInt(n: Int): UInt = UInt(n)
+    override def fromLong(n: Long): UInt = UInt(n.toInt)
+    override def fromFloat(n: Float): UInt = UInt(n.toLong.toInt)
+    override def fromDouble(n: Double): UInt = UInt(n.toLong.toInt)
+
+    override def createArray(data: Array[Int]): ValueClassScalarArray[UInt, Int] = ValueClassScalarArray(data)(this)
+    override def toUnderlying(u: UInt): Int = u.toInt
+    override def fromUnderlying(p: Int): UInt = UInt(p)
+
+  }
+
+  class ULongIsScalar extends ValueClassScalar[ULong, Long] {
+    override def toByte(a: ULong): Byte = a.toByte
+    override def toShort(a: ULong): Short = a.toShort
+    override def toInt(a: ULong): Int = a.toInt
+    override def toLong(a: ULong): Long = a.toLong
+    /* toFloat() and toDouble() seem to be (currently) broken in spire.
+     * Issue is filed here: https://github.com/non/spire/issues/387
+     * For now, we take the detour via BigInt.
+     */
+    override def toFloat(a: ULong): Float = a.toBigInt.toFloat
+    override def toDouble(a: ULong): Double = a.toBigInt.toDouble
+
+    override def fromByte(n: Byte): ULong = ULong(n.toLong)
+    override def fromShort(n: Short): ULong = ULong(n.toLong)
+    override def fromInt(n: Int): ULong = ULong(n.toLong)
+    override def fromLong(n: Long): ULong = ULong(n)
+
+    /* Float.toLong() and Double.toLong() have the annoying habit of
+     * capping their output to the range of Long. This is not an issue
+     * for negative numbers (the result is likely to be wrong anyway,
+     * as we don't expect negative numbers), but the ULong range is
+     * actually twice as large as Long's when it comes to positive values.
+     */
+    override def fromFloat(n: Float): ULong = {
+      val l = n.toLong
+      l match {
+        case Long.MaxValue => ULong(BigDecimal(n.toDouble).toLong)
+        case _ => ULong(l)
+      }
+    }
+    override def fromDouble(n: Double): ULong = {
+      val l = n.toLong
+      l match {
+        case Long.MaxValue => ULong(BigDecimal(n).toLong)
+        case _ => ULong(l)
+      }
+    }
+
+    override def createArray(data: Array[Long]): ValueClassScalarArray[ULong, Long] = ValueClassScalarArray(data)(this)
+    override def toUnderlying(u: ULong): Long = u.toLong
+    override def fromUnderlying(p: Long): ULong = ULong(p)
+  }
+}
+
+sealed trait ScalarArray[S] {
+  def apply(index: Int): S
+  def length: Int
+  final lazy val size = length
+
+  final def isDefinedAt(index: Int): Boolean = index < size && index >= 0
+
+  def map[T: Scalar: ClassTag](f: S => T): ScalarArray[T]
+
+  def iterator: Iterator[S]
+}
+
+abstract case class AbstractScalarArray[S, U](protected[scalismo] val rawData: Array[U]) extends ScalarArray[S] {
```

Comment:
Is this class intended to be used from outside? If not we should make it private. If yes, we should document it.